### PR TITLE
MMDiT Optimization

### PR DESCRIPTION
**Pull request by lapp0**
With these changes, we can comfortably fit `n_frames: 2000`, `batch_size: 64` into memory.

Runs (1x8 H200):
- `debug_mmdit`: https://wandb.ai/louiscastricato/video_models/runs/i80e5kby
- this PR: https://wandb.ai/louiscastricato/video_models/runs/jkinsz4m

### Changes

Optimizations:
- Increase dataloader throughput
- Log both video / audio loss
- Optimize `AVRoPE`
- Compile backbone

Bug fixes:
- fix name error: `MMDiTBlock2` -> `MMDiTBlock`

### CODE REVIEW

File: owl_wms/nn/mmattn.py [hunk 1]
```diff
@@ -274,20 +204,40 @@ def test_mask():
...
     n_frames = 10
-    n_tok_per_frame = 16
-    n_context = 16
+    n_tok_per_frame = 17
+    n_audio_tokens = 1
+    total_tokens = n_frames * n_tok_per_frame
 
-    mask = create_block_causal_mask_with_mm(n_frames*n_tok_per_frame, n_context, n_tok_per_frame)
+    # TODO !!!!
+    mask = create_block_causal_mask_with_mm(total_tokens, n_tok_per_frame, n_audio_tokens)
```
Comment (lapp0) [inline owl_wms/nn/mmattn.py:212]: TODO: update this test

File: owl_wms/nn/modulation.py [hunk 2]
```diff
@@ -41,4 +40,24 @@ def forward(self, x, cond):
...
+    x_norm = layer_norm(x)
+    return x_norm * (1 + scale) + bias
+
+def cond_gate(x, gate):
+    # gate: [b, n, d], x: [b, n*m, d]
+    b, nm, d = *x.shape[:2], x.size(-1)
+    n = gate.size(1)
+    m = nm // n
+    gate = gate.view(b, n, 1, d).expand(-1,-1,m,-1).reshape(b, nm, d)
+    return gate * x
```
Comment (lapp0) [inline owl_wms/nn/modulation.py:63]: Dit-Air

File: owl_wms/nn/normalization.py [hunk 3]
```diff
@@ -27,15 +30,14 @@ def forward(self, x):
...
-
     def forward(self, q, k):
-        return self.q_norm(q), self.k_norm(k)
+        return layer_norm(q), layer_norm(k)
 
 def LayerNorm(dim):
-    return nn.LayerNorm(dim, elementwise_affine = False)
\ No newline at end of file
+    # TODO: remove and replace with `def layer_norm` above
+    return nn.LayerNorm(dim, elementwise_affine = False)
```
Comment (lapp0) [inline owl_wms/nn/normalization.py:43]: Still in `attn.py`. `def layer_norm(...)` is equivalent assuming we're normalizing against the last dim.

File: owl_wms/utils/vis.py [hunk 4]
```diff
@@ -11,6 +11,7 @@ def draw_frame(frame, mouse, button):
     # frame is a torch tensor of shape [3,h,w]
     # mouse is [2,] tensor
     # button is list[bool]
+    frame = frame[:3] # Only ever take 3 channels
```
Comment (lapp0) [inline owl_wms/utils/vis.py:14]: TODO: review whether this is necessary

File: owl_wms/trainers/av_trainer.py [hunk 5]
```diff
@@ -136,12 +133,14 @@ def get_ema_core():
         timer = Timer()
         timer.reset()
         metrics = LogHelper()
-        if self.rank == 0:
-            wandb.watch(self.get_module(), log = 'all')
-        
+        #if self.rank == 0:
+        #    wandb.watch(self.get_module(), log = 'all')
```
Comment (lapp0) [inline owl_wms/trainers/av_trainer.py:137]: From upstream, probably want to keep this uncommented

File: owl_wms/nn/mmattn.py [hunk 6]
```diff
@@ -1,229 +1,159 @@
...
-        self.out_2 = nn.Linear(config.d_model, config.d_model)
+    def split_qkv(self, qkv, tok_per_frm):
+        return eo.rearrange(qkv, 'b (f n) (three h d) -> three b h f n d', n=tok_per_frm, three=3, h=self.n_heads)
 
-        self.qk_norm_1 = QKNorm(config.d_model // config.n_heads)
-        self.qk_norm_2 = QKNorm(config.d_model // config.n_heads)
+    def forward(self, x0, x1, block_mask=None, kv_cache=None):
+        """
+        For MMDiT we assume kv_cache is a tuple of two caches
+        """
```
Comment (lapp0) [inline owl_wms/nn/mmattn.py:48]: Stale, incorrect comment

**Review by shahbuland — APPROVED** [review]
LGTM
